engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -450,6 +522,62 @@ boolean equalsHashMaps(Object selfHashMap, Object otherHashMap,
     }
   }
 
+  @Specialization(guards = {
+      "!isAtom(selfObject)",
+      "!isAtom(otherObject)",
+      // Objects with types are handled in `equalsTypes` specialization, so we have to
+      // negate the guards of that specialization here - to make the specializations
+      // disjunctive.
```

Comment:
Correct. Disjunctive specializations have to have exclusive guards so the order of applying them doesn't matter.